### PR TITLE
Different approach to parsing/serving surrogate content

### PR DESCRIPTION
**Pull request by bsstoner**
**Reviewer:** @jdorweiler 

## Description:

This PR sets up the extension to detect outbound requests to GA and return data redirects to noop surrogate functions in their place.

As of now we're only addressing a handful of necessary cases from that list, but we've structured this so that we may add more without needing to change any code here.

## Steps to test this PR:
1. npm run dev-firefox (or chrome)
2. reload extension
3. in the console, check that surrogateList exists and is properly structured data
4. nuke whitelist `whitelists.trackersWhitelist.isLoaded = false`
5. visit broken site (I've been using raspberrypi.org, follow steps to buy)
6. request to GA should be intercepted, and a redirect to a base64 data url should be returned.
7. most importantly, the tracker is blocked and the page still functions.

You can further verify that the tracker was blocked by checking the network tab like this:
virtualbox_2018-01-12_18-11-23

## Automated tests:
- [x] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [x] Get advice or leverage existing code
- [x] Agree on technical approach with reviewer (if the changes are nuanced)
- [x] Ensure that there is a testing strategy (and documented non-automated tests)
- [x] Ensure there is a documented monitoring strategy (if necessary)
- [x] Consider systems implications

### CODE REVIEW

**Comment by dharb** [comment]
I just made a couple updates here:
1. use parser from https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/19
2. make sure we return the redirect url and not just block the offending tracker here: https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/21/files#diff-bce9949792f0ac5d954c3c70b4fd2563R209

I ran the surrogate and tracker unit tests, everything looks good there.

**Comment by bsstoner** [comment]
Getting a CSP error on some sites...not sure if we can easily do anything about that though.

screen shot 2018-01-15 at 10 05 19 am

File: shared/js/surrogates.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        //
+        // Anything not specific in the rule is intended to be a wildcard, including the paths.
+        //
+        // So that rule can match things like:
+        // https://wwww.googletagservices.com/js/gpt.js
+        // or
+        // http://en.www.googletagservices.com/some/other/random/path/gpt.js?v=123
+        //
+        // All our rules have domain + filename, so for now we're safe making that assumption.
+        let splitUrl = url.split('/')
```
Comment (jdorweiler) [inline shared/js/surrogates.js]: We're loading [tldjs](https://www.npmjs.com/package/tldjs) in the extension. Did you try using that?

Comment (bsstoner) [reply shared/js/surrogates.js]: yea, I'm using that to get the domain, but I didn't think it had functionality to parse path/filename?

Comment (jdorweiler) [reply shared/js/surrogates.js]: Yeah you're right. Looks like it doesn't include the path or file.

**Comment by bsstoner** [comment]
Re: CSP headers, as discussed in O-1 meeting, uBlock has the same issue, so we're not going to worry about it for now.

File: shared/js/background.js [hunk 2]
```diff
@@ -200,8 +201,14 @@ chrome.webRequest.onBeforeRequest.addListener(
                     console.info( "blocked " + utils.extractHostFromURL(thisTab.url)
                                  + " [" + tracker.parentCompany + "] " + requestData.url);
 
+                    // return surrogate redirect if match, otherwise
                     // tell Chrome to cancel this webrequest
-                    return {cancel: true};
+                    if (tracker.redirectUrl) {
+                        console.info("serving surrogate content for: ", tracker.url)
```
Comment (jdorweiler) [inline shared/js/background.js]: Can you remove this message? It's the same as the one coming`checkSurrogateList` in trackers.js

Comment (bsstoner) [reply shared/js/background.js]: 👍  done

**Comment by jdorweiler** [comment]
👍

**Pull request merged**